internal/states/encryption/encryption.go

```diff
@@ -0,0 +1,271 @@
+			key,
+		}] = config
+	}
+
+	return nil
+}
+
+func (e *encryption) ApplyHCLEncryptionConfiguration(key encryptionconfig.Key, config encryptionconfig.Config) error {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	if err := key.Validate(); err != nil {
+		return fmt.Errorf("failed to parse encryption configuration from HCL (%w)", err)
+	}
+	meta := encryptionconfig.Meta{
+		encryptionconfig.SourceHCL,
+		key,
+	}
+	e.encryptionConfigs[meta] = config
+	return nil
+}
+
+func (e *encryption) ApplyHCLDecryptionFallbackConfiguration(key encryptionconfig.Key, config encryptionconfig.Config) error {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	if err := key.Validate(); err != nil {
+		return fmt.Errorf("failed to parse decryption fallback configuration from HCL (%w)", err)
+	}
+	meta := encryptionconfig.Meta{
+		encryptionconfig.SourceHCL,
+		key,
+	}
+	e.decryptionFallbackConfigs[meta] = config
+	return nil
+}
+
+func (e *encryption) Validate() (diags tfdiags.Diagnostics) {
+	keys := e.allConfigKeys()
+
+	for key := range keys {
+		if _, err := e.build(key); err != nil {
+			diags = append(diags, tfdiags.Sourceless(
+				tfdiags.Error,
+				fmt.Sprintf("Invalid state encryption configuration for configuration key %s", key),
+				err.Error(),
+			))
+		}
+	}
+
+	return diags
+}
+
+func (e *encryption) allConfigKeys() map[encryptionconfig.Key]struct{} {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	// Collect all the keys between the encryption and decryption fallback configuration. We do this to make sure
+	// we don't process keys twice, which would result in duplicate errors. We also make sure that the
+	// default key is tested as a backend key since that's where it will actually be used.
+	keys := map[encryptionconfig.Key]struct{}{}
+	addKey := func(key encryptionconfig.Key) {
+		if key == encryptionconfig.KeyDefaultRemote {
+			// The presence of the default key means that there should be a valid backend configuration as the
+			// default key is automatically applied to it.
+			keys[encryptionconfig.KeyBackend] = struct{}{}
+		} else {
+			keys[key] = struct{}{}
+		}
+	}
+	for meta := range e.encryptionConfigs {
+		addKey(meta.Key)
+	}
+	for meta := range e.decryptionFallbackConfigs {
+		addKey(meta.Key)
+	}
+
+	return keys
+}
+
+func (e *encryption) RemoteState() (encryptionflow.StateFlow, error) {
+	return e.build(encryptionconfig.KeyBackend)
+}
+
+func (e *encryption) StateFile() (encryptionflow.StateFlow, error) {
+	return e.build(encryptionconfig.KeyStateFile)
+}
+
+func (e *encryption) PlanFile() (encryptionflow.PlanFlow, error) {
+	return e.build(encryptionconfig.KeyPlanFile)
+}
+
+func (e *encryption) RemoteStateDatasource(configKey encryptionconfig.Key) (encryptionflow.StateFlow, error) {
+	if !configKey.IsRemoteDataSource() {
+		return nil, fmt.Errorf("the specified configuration key is not a valid remote data source key (this is likely a bug, did you want to call RemoteState(), StateFile(), or PlanFile()?)")
```

Comment:
could this be a panic? There's a chance that if this is an implementation error that it shouldn't be ignorable.

Reply:
This can also be a configuration error if we let users specify this, depending on how the calling side is implemented.